homeassistant/components/hue/binary_sensor.py

```diff
@@ -34,6 +34,10 @@ def is_on(self):
     def device_state_attributes(self):
         """Return the device state attributes."""
         attributes = super().device_state_attributes
+        if "on" in self.sensor.config:
+            attributes["on"] = self.sensor.config["on"]
+        if "reachable" in self.sensor.config:
+            attributes["reachable"] = self.sensor.config["reachable"]
```

Comment:
We should not do this. This is already represented as `available`. And `on` should also become part of that. If a device is not on, it should be marked as unavailable.

Reply:
Hi balloob
while I appreciate the remark regarding reachable here (I added it merely because it is available and it can be useful to have that separate info in core HA) we really need the 'on' attribute, because the binary_sensors are not only sensors in the Hue ecosystem, but also switches with a on/off state. That state is currently not available in core HA.

using switches like this:
```
      laundry_sensor_motion_switch:
        friendly_name: 'Laundry motion sensor switch'
        command_on: >
          curl -X PUT -d '{"on":true}' "http://192.168.1.212/api/dmredactedjSYxaxM9pQq/sensors/22/config/"
        command_off: >
          curl -X PUT -d '{"on":false}' "http://192.168.1.212/api/dmredactedpojSYxaxM9pQq/sensors/22/config/"
        command_state: !secret get_laundry_config
        value_template: >
          {{value_json.config.on}}
```

makes clear why I hope.

Since this is a very small addition to the core binary_sensor.py, I ask you to reconsider what you said above.
thanks.

Reply:
> And `on` should also become part of that. If a device is not on, it should be marked as unavailable.

Hi Paulus, this is truly incorrect.. 
If the binary_sensor is not 'on', it is what it says: not 'on'. So 'off' ;-)  In which state it doesn't sense motion. But it is reachable and very much available.

This has nothing to do with being unavailable, this is the state of the binary_sensor (switch) being turned on /off (either in the app or by other means)

check below: the sensor has state 'off' meaning no motion is detected. It also has attribute.on: true, meaning the sensor is turned 'on'.
this is vital information now missing from the Hue core integration.

<img width="760" alt="Schermafbeelding 2020-03-10 om 08 16 17" src="https://user-images.githubusercontent.com/33354141/76288622-729a5980-62a7-11ea-8a1a-c3240b98f5dc.png">

Please reconsider adding this to the binary_sensor attributes

Reply:
I agree with balloob in this case.

The reachable state should be reflected as "unavailable". If it doesn't do that correctly, that should be considered a bug.

> I added it merely because it is available and it can be useful to have that separate info in core HA

That is not what we doing with Home Assistant. Entities should reflect the actual device/entity states (translated to the HA world) and not come up with their own.

> we really need the 'on' attribute, because the binary_sensors are not only sensors in the Hue ecosystem

A `binary_sensor` has 2 states. True & false, which can be represented according to the device type; resulting in, for example, on/off. Adding in another "on" state, can be an indicator for either of the following:

- It should have been another entity.
- The state of device isn't properly reflected in the current binary sensor.
- It should not have been a binary sensor.

Please note, it doesn't mean the current implementation is wrong or should be resolved using one of the above.

So looking at the case you are raising for the motion sensor, if I understand it correctly, the motion sensor's motion-sensing capabilities can be turned on/off. IMHO, that is a switch in the HA world.

In all the above cases, the addition of both, doesn't fit in the paradigm of Home Assistant.

Reply:
> So looking at the case you are raising for the motion sensor, if I understand it correctly, the motion sensor's motion-sensing capabilities can be turned on/off. IMHO, that is a switch in the HA world.

HI Frenck,
thanks for chiming in here.
yes, that is true, and why people need to create these now (remember that most people have quite a few of these, 12 here, so it takes a lot of extra code and Hub communication :

```
      attic_motion_sensor_switch:
        friendly_name: 'Attic motion sensor switch'
        command_on: >
          curl -X PUT -d '{"on":true}' "http://192.168.1.212/api/dredacted9pQq/sensors/12/config/"
        command_off: >
          curl -X PUT -d '{"on":false}' "http://192.168.1.212/api/dredacted9pQq/sensors/12/config/"
        command_state: curl http://192.168.1.212/api/dredacted9pQq/sensors/12/
        value_template: >
          {{value_json.config.on}}
```

thing is we keep needing extra rest sensors for these values, while the Hue core integration could very easily offer them without any extra Hub stress. Since that is a major issue in Hue/HA on a bigger scale, relieving the Hub from any unnecessary stress should be in the paradigm of Home Assistant most certainly.

I fully understand And appreciate what you're saying, and yes, it might have been better to natively offer the motion_sensor_switch (please don't hold back ;-) )
Since that however isn't the case, adding this minute 'on' attribute could solve it very quickly.

on the reachable: I agree in the case of the motion_sensors this is less relevant, and not_reachable could be considered 'unavailable', though that really isn't the case.

------
on lights however, 'reachable' really is another beast all together. Where unavailable is signaling a Hue communication issue, in regular operation lights can truly be unreachable, because they are cut off power. 
With the allow_unreachable set to true, these lights simply show on/off while they in fact are Off power.  Without that, all power cut lights would show unavailable...which is an unsatisfactory generalization. We need to be able to differentiate between unavailable and unreachable. And we can, because the Hue API offers that, and HA can very easily use that available info.

1 picture says all:

<img width="385" alt="Schermafbeelding 2020-03-10 om 09 58 59" src="https://user-images.githubusercontent.com/33354141/76296003-d166cf80-62b5-11ea-83bb-569f2bde71a4.png">

the Garden terrace light isn't 'unavailable' (doesn't have an error communicating with the Hub) but is unreachable (cut off power purposely). As you can see the Hub still says it has brightness 50%, which is correct on the Hub, but not on the light. It is not connected to power...)

again we now need yet again extra rest/command sensors a la:
```
binary_sensor:
  - platform: command_line
    name: Garden backyard reachable
    command: !secret garden_backyard_command
    value_template: >
      {{value_json.state.reachable}}
    payload_on: 'True'
    payload_off: 'False'
    device_class: light
```

or 
```
  - platform: rest
    name: Hue lights config
    resource: !secret hue_lights_resource
    method: GET
    value_template: >
      {{value_json|length}}
    json_attributes:
      - '10' #mobile
      - '13' #28951 
      - '16' #garden_terrace
      - '17' #garden_backyard
      - '41' #gate_outdoors
      - '42' #porch_outdoors
      - '50' #bureau_marte
      - '53' #backdoor_outdoors
    scan_interval: 300
```
while this all could be simply in Core HA/Hue...